### PR TITLE
4 ➡️ 5

### PR DESCRIPTION
**Pull request by mjcarroll**
# ➡️ Forward port

Port ign-common4 to gz-common5

Branch comparison: https://github.com/gazebosim/gz-common/compare/gz-common5...ign-common4

**Note to maintainers**: Remember to **Merge** with commit (not squash-merge or rebase)

### CODE REVIEW

**Comment by nkoenig** [comment]
@mjcarroll, when can the plugin functionality in `gz-common` be removed?

**Comment by mjcarroll** [comment]
In main/6, I have already removed it. Since it is deprecated in 5 it is left out of the gz headers and is only in ignition.

**Comment by nkoenig** [comment]
@mjcarroll looks like the ABI checker had some trouble.

**Comment by mjcarroll** [comment]
> @mjcarroll looks like the ABI checker had some trouble.

Also added some compiler warnings, looking into it.

**Comment by mjcarroll** [comment]
This PR https://github.com/gazebosim/gz-common/pull/427 needed to be ported over to protect the deprecated headers as well

**Comment by mjcarroll** [comment]
ABI Checker is complaining about:

```
TempDirectory.hh
namespace gz::common
TempDirectory::__ct [C1] ( TempDirectory const& p1 )
_ZN2gz6common13TempDirectoryC1ERKS1_

TempDirectory::__ct [C2] ( TempDirectory const& p1 )
_ZN2gz6common13TempDirectoryC2ERKS1_
```

Which I believe is a false positive?  The temp directory uses the UniqueImplPtr pattern (and has for a while), so I believe that the copy constructors should have been deleted by default?

**Comment by nkoenig** [comment]
> ABI Checker is complaining about:
> 
> ```
> TempDirectory.hh
> namespace gz::common
> TempDirectory::__ct [C1] ( TempDirectory const& p1 )
> _ZN2gz6common13TempDirectoryC1ERKS1_
> 
> TempDirectory::__ct [C2] ( TempDirectory const& p1 )
> _ZN2gz6common13TempDirectoryC2ERKS1_
> ```
> 
> Which I believe is a false positive? The temp directory uses the UniqueImplPtr pattern (and has for a while), so I believe that the copy constructors should have been deleted by default?

In `ign-common4`, `TempDirectory` uses `std::unique_ptr<TempDirectoryPrivate> dataPtr`. In `gz-common5`, `TempDirectory` uses `GZ_UTILS_UNIQUE_IMPL_PTR`. This could be causing the confusion.

I think we can safely ignore this ABI failure.

File: av/include/ignition/common/HWEncoder.hh [hunk 1]
```diff
@@ -1,19 +0,0 @@
...
- *
- * Unless required by applicable law or agreed to in writing, software
- * distributed under the License is distributed on an "AS IS" BASIS,
- * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
- * See the License for the specific language governing permissions and
- * limitations under the License.
- *
- */
-
-#include <gz/common/HWEncoder.hh>
```
Comment (scpeters) [inline av/include/ignition/common/HWEncoder.hh:18]: why is this file deleted?

File: include/gz/common/Console.hh [hunk 2]
```diff
@@ -31,8 +31,6 @@ namespace gz
 {
   namespace common
   {
-    // TODO(CH3): Deprecated. Remove all the ign macros on tock.
-    /// \brief Output an error message, if the verbose level is >= 1
```
Comment (scpeters) [inline include/gz/common/Console.hh:35]: should this comment be deleted?

File: include/ignition/common/SpecializedPlugin.hh [hunk 3]
```diff
@@ -14,8 +14,6 @@
  * limitations under the License.
  *
  */
-#ifndef IGNITION_COMMON_SPECIALIZEDPLUGIN_HH_
-#define IGNITION_COMMON_SPECIALIZEDPLUGIN_HH_
```
Comment (scpeters) [inline include/ignition/common/SpecializedPlugin.hh:18]: should these header guards remain?

File: include/ignition/common/detail/SpecializedPlugin.hh [hunk 4]
```diff
@@ -19,7 +19,6 @@
 #define IGNITION_COMMON_DETAIL_SPECIALIZEDPLUGIN_HH_
 
 #include <ignition/common/config.hh>
-#include "ignition/common/SpecializedPlugin.hh"
```
Comment (scpeters) [inline include/ignition/common/detail/SpecializedPlugin.hh:22]: I think this can be retained if its header guards are restored

File: profiler/include/CMakeLists.txt [hunk 5]
```diff
@@ -0,0 +1,2 @@
+install(DIRECTORY gz DESTINATION ${IGN_INCLUDE_INSTALL_DIR_FULL})
+install(DIRECTORY ignition DESTINATION ${IGN_INCLUDE_INSTALL_DIR_FULL})
```
Comment (scpeters) [inline profiler/include/CMakeLists.txt:2]: this file can be deleted since these commands are in `profiler/CMakeLists.txt`

File: av/include/ignition/common/HWEncoder.hh (hunk 1 above)
Comment (mjcarroll) [reply av/include/ignition/common/HWEncoder.hh:18]: `HWEncoder.hh` is a private header, and is prevented from installation in the CMakeLists here: https://github.com/gazebosim/gz-common/blob/b9859bdfb5f7db2839133b10ae7a7a3dfc6eadb9/av/include/gz/common/CMakeLists.txt#L1-L2

I think we don't need a corresponding ignition/common header for this reason.

File: include/gz/common/Console.hh (hunk 2 above)
Comment (mjcarroll) [reply include/gz/common/Console.hh:35]: Nope, restored.

File: include/ignition/common/SpecializedPlugin.hh (hunk 3 above)
Comment (mjcarroll) [reply include/ignition/common/SpecializedPlugin.hh]: 54cbbac4

File: include/ignition/common/detail/SpecializedPlugin.hh (hunk 4 above)
Comment (mjcarroll) [reply include/ignition/common/detail/SpecializedPlugin.hh]: 54cbbac4

File: profiler/include/CMakeLists.txt (hunk 5 above)
Comment (mjcarroll) [reply profiler/include/CMakeLists.txt]: 54cbbac4

File: av/include/ignition/common/HWEncoder.hh (hunk 1 above)
Comment (scpeters) [reply av/include/ignition/common/HWEncoder.hh:18]: gotcha; this makes sense. In sdformat, we've moved non-installed headers to the `src` folder so that all headers in an `include` folder are installed, so I didn't think to check if it was excluded. No changes needed

**Review by scpeters — APPROVED** [review]
